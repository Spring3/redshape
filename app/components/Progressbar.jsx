import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

const Background = styled.div`
  background: ${props => props.theme.bgLight};
  position: relative;
  border-radius: 5px;
  height: ${props => props.height}px;
`;

const Progress = styled.div`
  border-radius: 5px;
  max-width: 100%;
  width: 0;
  ${props => css`
    transition: width ease ${props.transitionTime};
    width: ${props.percent}% !important;
    height: ${props.height}px;
    background: ${props.background};
  `}
`;

const Progressbar = ({ percent, background, id, className, height }) => {
  const percentage = (isFinite(percent) && !isNaN(percent)) ? percent : 0;
  return (
    <Wrapper
      id={id}
      className={className}
    >
      <Background
        height={height}
      >
        <Progress
          percent={percentage}
          background={background}
          height={height}
        />
      </Background>
    </Wrapper>
  );
};

Progressbar.propTypes = {
  percent: PropTypes.number.isRequired,
  background: PropTypes.string.isRequired,
  className: PropTypes.string,
  id: PropTypes.string,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ])
};

Progressbar.defaultProps = {
  className: undefined,
  id: undefined,
  height: 5
};

export default Progressbar;
