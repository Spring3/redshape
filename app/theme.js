export default {
  bg: '#FFFFFF',
  bgLight: '#F9F9F9',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#909090',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  transitionTime: '.2s'
};
