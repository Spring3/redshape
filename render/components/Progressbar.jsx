import React from 'react';
import styled, { withTheme, css } from 'styled-components';
import PropTypes from 'prop-types';

import Tooltip from './Tooltip';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

const Background = styled.div`
  background: ${(props) => props.theme.bgDark};
  position: relative;
  border-radius: 5px;
  height: ${(props) => props.height}px;
`;

export const Progress = styled.div`
  border-radius: 5px;
  max-width: 100%;
  width: 0;
  ${(props) => css`
    float: ${props.float || 'left'};
    transition: width ease ${props.transitionTime};
    width: ${props.percent}% !important;
    height: ${props.height}px;
    background: ${props.background};
  `}
`;

const Progressbar = ({
  percent, background, id, className, height, mode, theme
}) => {
  let percentage = (isFinite(percent) && !isNaN(percent)) ? percent : 0;
  let percentageOver;
  const percentageText = `${percentage.toFixed(0)}%`;
  if (mode === 'progress-gradient') {
    const colors = ['red', 'yellow-red', 'yellow', 'yellow-green', 'green'];
    const ranges = [20, 40, 60, 80, 100];
    const colorIdx = ranges.findIndex((el) => el >= percentage);
    const color = colors[colorIdx] || 'green';
    background = theme[color];
  } else if (mode === 'time-tracking') {
    background = theme.green;
    if (percentage >= 75.0) {
      background = theme['yellow-green'];
    }
    if (percentage >= 100.0) {
      percentageOver = percentage - 100;
      percentageOver = (percentageOver / percentage) * 100;
      percentage = (100 / percentage) * 100;
    }
  }

  return (
    <Wrapper
      id={id}
      className={className}
    >
      <Tooltip text={percentageText}>
        <Background
          height={height}
        >
          <Progress
            percent={percentage}
            background={background}
            height={height}
          />
          {
          percentageOver > 0 && (
            <Progress
              float="right"
              percent={percentageOver}
              background={theme.red}
              height={height}
            />
          )
        }
        </Background>
      </Tooltip>
    </Wrapper>
  );
};

Progressbar.propTypes = {
  percent: PropTypes.number.isRequired,
  background: PropTypes.string.isRequired,
  className: PropTypes.string,
  id: PropTypes.string,
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  mode: PropTypes.string
};

Progressbar.defaultProps = {
  className: undefined,
  id: undefined,
  height: 5,
  mode: 'default'
};

export default withTheme(Progressbar);
