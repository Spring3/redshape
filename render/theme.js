export default {
  bg: '#FFFFFF',
  bgDark: '#F9F9F9',
  bgDarker: '#E9E9E9',
  bgDisabled: 'hsl(0,0%,95%)',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#A0A0A0',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  shadow: '#D0D0D0',
  transitionTime: '.2s'
};
