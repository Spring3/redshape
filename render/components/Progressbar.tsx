import React from 'react';
import { useTheme } from 'styled-components';
import { css } from '@emotion/react';

import { Tooltip } from './Tooltip';
import { Flex } from './Flex';
import { theme as Theme } from '../theme';

type ProgressbarProps = {
  percent?: number;
  className?: string;
}

const Progressbar = ({
  percent = 0, className
}: ProgressbarProps) => {
  const theme = useTheme() as typeof Theme;
  // eslint-disable-next-line
  const percentage = (isFinite(percent) && !isNaN(percent)) ? percent : 0;
  const percentageText = `${percentage.toFixed(0)}%`;

  return (
    <Flex
      direction="column"
      justifyContent='center'
      alignItems='center'
      className={className}
    >
      <Tooltip text={percentageText}>
        <div
          css={css`
            background: ${theme.bgDark};
            position: relative;
            border-radius: 5px;
            height: 5px;
          `}
        >
          <div
            css={css`
              border-radius: 5px;
              max-width: 100%;
              width: 0;
              transition: width ease ${theme.transitionTime};
              width: ${percent}% !important;
              height: 5px;
              background: ${theme.mainDark};
            `}
          />
        </div>
      </Tooltip>
    </Flex>
  );
};

export {
  Progressbar
};
