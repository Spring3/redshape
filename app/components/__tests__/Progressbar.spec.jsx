import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Progressbar from '../Progressbar';
import theme from '../../theme';

describe('Progressbar component', () => {
  it('should match the snapshot', () => {
    const tree = renderer.create(
      <Progressbar percent={25} background="yellow" height={10} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('should fallback in case infinite number was given as percentage', () => {
    const wrapper = shallow(
      <div>
        <Progressbar theme={theme} percent={Infinity} />
        <Progressbar theme={theme} percent={NaN} />
      </div>
    );

    wrapper.find(Progressbar).forEach(node => expect(node.dive().childAt(0).childAt(0).prop('percent')).toBe(0));
  });

  it('should allow the height and background to be customized', () => {
    const wrapper = shallow(
      <Progressbar theme={theme} height={20} background="salmon" />
    );
    expect(wrapper.dive().childAt(0).prop('height')).toBe(20);
    expect(wrapper.dive().childAt(0).childAt(0).prop('background')).toBe('salmon');
    expect(wrapper.dive().childAt(0).childAt(0).prop('height')).toBe(20);
  });
});
