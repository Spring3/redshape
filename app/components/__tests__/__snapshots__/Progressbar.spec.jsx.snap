// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progressbar component should match the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  border-radius: 5px;
  height: 10px;
}

.c2 {
  border-radius: 5px;
  max-width: 100%;
  width: 0;
  -webkit-transition: width ease;
  transition: width ease;
  width: 25% !important;
  height: 10px;
  background: yellow;
}

<div
  className="c0"
>
  <div
    className="c1"
    height={10}
  >
    <div
      className="c2"
      height={10}
    />
  </div>
</div>
`;
